docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,541 @@
+- We keep the same behavior as in the previous implementation, where parameters
+  are updated by the last optimizer in the sequence, in the case where a path
+  matches more than one `FilterFn`.
+- Note that the signature in `Optimizer` from the [functional API] needs to be
+  changed slightly to accept and return
+  `Union[OptimizerState, Sequence[OptimizerState]]` to make this work.
+
+## Functional Usage
+[Functional Usage]: #functional-usage
+
+Using the [functional API] directly makes for the following code:
+
+```python
+def make_train_step(apply_fn, opt_update_fn):
+  @jax.jit
+  def train_step(opt_state, variables, inputs, labels):
+
+    def loss_fn(params):
+      logits, new_model_state = apply_fn(
+          variables.copy(dict(params=params)), inputs, mutable=['batch_stats'])
+      loss = xent_loss(logits, labels)
+      return loss, new_model_state
+
+    variables, params = variables.pop('params')
+    (loss, new_model_state), grads = jax.value_and_grad(loss_fn, has_aux=True)(
+        params)
+    new_params, new_opt_state = opt_update_fn(
+        grads, params, opt_state)
+    new_variables = variables.copy(dict(params=new_params, **new_model_state))
+    return new_opt_state, new_variables, loss
+
+  return train_step
+
+opt_state = optim.init(variables['params'])
+train_step = make_train_step(model.apply, optim.update)
+for batch in ds.as_numpy_iterator():
+  opt_state, variables, loss = train_step(
+      opt_state, variables, batch['image'], batch['label'])
+```
+
+Remarks:
+
+- Compared with the [previous API], we can now keep the entire `variables`
+  including the `params` as an input and output to the `train_step()`.
+- Splitting `params` from `variables` is still necessary inside the train step
+  because we only want to compute gradients with respect to `params` and not the
+  entire `variables`.
+- A further complication is that we cannot provide `apply_fn` and
+  `opt_update_fn` as direct arguments to the jitted function `train_step()`
+  because they're not valid JAX types - they need to be wrapped in a
+  `make_train_step()` function like above, or declared as `static_argnums` to
+  `jax.jit()`.
+
+## Linen Helper
+[Linen helper]: #linen-helper
+
+While the above training loop is perfectly fine, a number of lines are going to
+be exactly the same for every Linen training loop, namely:
+
+1. The splitting of variables.
+2. Passing in parameters separately because we only need the gradients for
+   those.
+3. Updating the parameters with the opimizer.
+4. Merging the updated parameters, any other updated collections, and the
+   unchanged collections to generate the new variables.
+5. We also need to keep track of the model's `apply_fn`, the otpimizer's
+   `update_fn`, and the optimizer state, in addition to the data.
+
+
+If we put all this state into a dataclass then we only have to pass around a
+single object (and the train data):
+
+```python
+@flax.struct.dataclass
+class LinenHelper:
```

Comment:
`class LinenHelper(PyTreeNode)` is better for pytype

Reply:
Ah right!